src/Chapter3.hs

```diff
@@ -374,6 +382,34 @@ after the fight. The battle has the following possible outcomes:
 
 -}
 
+-- I decided to add a `1` prefix to these data structures because I repeated them in the last excercise of this chapter
+-- also, the two data structures could also be unified, using a phantom type parameter to distinguish them
+data Knight1 = Knight1 {
+  knight1Health :: Int,
+  knight1Attack :: Int,
+  knight1Gold :: Int
+} deriving Show
+
+data Monster1 = Monster1 {
+  monster1Health :: Int,
+  monster1Attack :: Int,
+  monster1Gold :: Int
+} deriving Show
+
+fight :: Knight1 -> Monster1 -> Int
+fight knight@(Knight1 kHealth kAttack kGold) monster@(Monster1 mHealth mAttack mGold) =  -- RecordWildcards, maybe?
+  if monster1Health newMonster > 0
+    then
+      if knight1Health newKnight <= 0
+      then -1
+      else kGold
+  else
+    kGold + mGold
+  where
+    newMonster = monster { monster1Health = mHealth - kAttack }
+    newKnight = knight { knight1Health = kHealth - mAttack }
+-- Not really happy with the implementation. Maybe there's a simpler/more elegant variant
```

Comment:
I see that these two values `newMonster` and `newKnight` are not used in the body. So, probably they are safe to remove :recycle:

Reply:
I use them in the lines above. I found it more readable than inlining them. Also, in the version with guards I use `newMonster` twice